tests/integration/index/query_with_relation_filter_test.go

```diff
@@ -0,0 +1,310 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package index
+
+import (
+	"testing"
+
+	testUtils "github.com/sourcenetwork/defradb/tests/integration"
+)
+
+func TestQueryWithIndexOnOneToManyRelation_IfFilterOnIndexedRelation_ShouldFilter(t *testing.T) {
+	req1 := `query {
+		User(filter: {
+			devices: {model: {_eq: "MacBook Pro"}}
+		}) {
+			name
+		}
+	}`
+	req2 := `query {
+		User(filter: {
+			devices: {model: {_eq: "iPhone 10"}}
+		}) {
+			name
+		}
+	}`
+	test := testUtils.TestCase{
+		Description: "Filter on indexed relation field in 1-N relation",
+		Actions: []any{
+			createSchemaWithDocs(`
```

Comment:
todo: I do think this is a bad use-case for auto-generating docs.  This tests (and many others here) are testing a very specific set of conditions and results, and really do care about what the initial state is.  At the moment, in this test (and others) I cannot see what that initial state is and if I want to know I have to spend considerable effort figuring that out.

Please change these, so that what state is being tested is very obvious to anyone reading.

Reply:
I strongly disagree. The initial state is put in a separate file you don't need to familiarise yourself with it every time you read another test case, as it's the same. Hence you do it once.
The tests with indexes (especially with relations) require a dozen (or better couple) of docs.
Soon I will be adding even more complex scenarios with one-to-many-to-many and other relations for which I think at least 40 docs would make sense. The test setup will take around 300 lines which I don't think anyone will have fun to read. Moreover, no developer wants to manually maintain all dockeys and relations.

Reply:
> as it's the same

It is the same for a handful of index-explain tests.  If people only had to look at these tests that would be fine, but we have many other tests to look at too.  I'd really rather not have to learn stuff like this every time I enter a new set/sub-framework of integration tests.

It's not like you need hundreds of them here.  You are asserting that 3 very specific documents are being returned, and reading the test I have no idea whether they exist, only they exist, what their other field values are etc.  I can't even see the values that the filter is acting on.

Reply:
> Moreover, no developer wants to manually maintain all dockeys and relations.

So improve the way we do that, we already have very similar logic in the test framework for handling this for mutations.  Don't hide the important stuff along with the dockeys.

Reply:
> You are asserting that 3 very specific documents are being returned

This doesn't mean that the initial state needs only 3 docs. These 3 docs need to be selected from others, meaning it makes sense to have at least 5, each of them has related docs (1-N relation), which means on average 2 sub docs for every doc (to make index testing meaningful). 

Therefore roughly 15 docs would make sense to the test. It's not fun to deal with this ungrateful routine work. And having several of such test, would make a developer (at least) almost blind to the differences between them.

Reply:
> dog: 2  // the doc index of Islam's dog  

it helps only a bit, there is still a lot of bookkeeping to do, and large test bodies are not very friendly.

Reply:
> It's not fun to deal with this ungrateful routine work

No it is not, but it is well worth a bit of upfront tedium to avoid future misery for maintainers later.

My problem is that if this test fails, potentially when working on an area unspecific to indexes, when the developer looks at this test instead of seeing the minimized set of documents required for this test, they have to instead navigate through `createSchemaWithDocs` (a very dense file), then to `docs.go`, and then they are presented with a massive dataset containing fields and documents that they really do not care about.  

Worse, they are not free to change anything in this file, as it will change X number of other tests that they may never have seen before and have no desire to change.  This is a very significant liability in the codebase.  Changing anything in `docs.go` is extremely unsafe.

Reply:
I think we need better utilities for generating large sets of documents with specific properties or relationships.

I worked on something like this a long time ago when doing the initial benchmark suite, that did exactly this (although I'm not sure the code is a good enough state to be reused).

The crux here is that the tests don't care about The specifics of the values of the docs, only that they exist for numeric reasons (eg: number of fetches). So generating is a better route than explicitly writing it out.

However, based on my previous work, i know this to be a largish undertaking to create that kind of framework. 

I think it's important to have this kind of utility, but I think it's too large in scope to put in this PR.

I think we can define the ergonomics and design of such a framework and complete it after the PR, and leave this as is if we agree it's a temp measure.

Reply:
> The crux here is that the tests don't care about The specifics of the values of the docs

These do.  They are filtering on specific values, and asserting that specific values are returned.  I am not against us having something like you described, or what Islam has built, but I am very much against it's use here.

Reply:
> They are filtering on specific values, and asserting that specific values are returned

well, it's hard to assert on @explain results if you don't know exactly how many documents you are going to fetch and filter. Explain result here is the main indicator that the indexes are working and values are irrelevant there.
And the assertion on logical output is there just to make sure it's still working as expected, and partly to show why explain results are what they are.

> I think we can define the ergonomics and design of such a framework and complete it after the PR, and leave this as is if we agree it's a temp measure.

I also was thinking about making a framework out of it, but it doesn't cover yet some relation cases. So I might take care of this later.

Reply:
I've created a task https://github.com/sourcenetwork/defradb/issues/1962
I might have missed few details, so feel free to edit it

Reply:
LGTM, cheers Islam :)